### PR TITLE
UNI-242: Optimize database key schema representing file, metric and data hierarchies #send-for-review

### PR DESCRIPTION
**Pull request by lscheinkman**
@brev Please review.
This PR covers database part of deleting Metrics and it's dependent data.
The connection to the UI will be in a separate PR.

### CODE REVIEW

File: unicorn/js/main/DatabaseService.js [hunk 1]
```diff
@@ -57,13 +56,14 @@ function DatabaseService(path) {
...
-  // this.validator.addSchema(AddressSchema, '/Address');
-
   this.levelup = levelup(location, {
     db: leveldown,
     valueEncoding: 'json'
-  })
-  this.dbh = sublevel(this.levelup);
+  });
+  this._root = sublevel(this.levelup);
+  this._files = this._root.sublevel('File');
```
Comment (brev) [inline unicorn/js/main/DatabaseService.js]: nice idea! thanks

File: unicorn/js/main/DatabaseService.js [hunk 2]
```diff
@@ -211,20 +188,17 @@ DatabaseService.prototype.putFile = function (file, callback) {
  * @param {Function} callback - Async result handler: function (error, results)
  */
 DatabaseService.prototype.putFileBatch = function (files, callback) {
-  let ops = [];
-  const table = this.dbh.sublevel('file');
-
   // validate
-  files.forEach((file) => {
```
Comment (brev) [inline unicorn/js/main/DatabaseService.js]: you can get the array index through the `forEach()` callback if you want too:

File: unicorn/js/main/DatabaseService.js [hunk 3]
```diff
@@ -259,20 +232,17 @@ DatabaseService.prototype.putMetric = function (metric, callback) {
...
  */
 DatabaseService.prototype.putMetricBatch = function (metrics, callback) {
-  const table = this.dbh.sublevel('metric');
-  let ops = [];
-
   // validate
-  metrics.forEach((metric) => {
-    const validation = this.validator.validate (metric, MetricSchema);
+  for (let i=0; i<metrics.length; i++) {
+    const validation = this.validator.validate (metrics[i], MetricSchema);
```
Comment (brev) [inline unicorn/js/main/DatabaseService.js]: extra space ".validate ("

**Comment by brev** [comment]
LGTM :+1: thanks

File: unicorn/js/main/DatabaseService.js (hunk 2 above)
Comment (lscheinkman) [inline unicorn/js/main/DatabaseService.js]: The problem with `forEach` is the ability to stop the loop on errors as in line [221](https://github.com/numenta/numenta-apps/pull/496/files#diff-12ed1b4007e330812796976e7dc66a1fL221)
See usage note in https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach

**Pull request merged**